contents/scala/Scala.md

```diff
@@ -0,0 +1,241 @@
+  }
+}
+
+object Sample {
+  def main(args: Array[String]): Unit = {
+    val example = new Sample with TraitA with TraitB with TraitC
+   example.print()
+  }
+}
+```
+The following is the execution result of above code:
+```
+print in TraitC
+print in TraitB
+print in TraitA
+print in Sample
+```
+The explanation is that the call to `print()` firstly executes the code in `TraitC`, which is the last trait mixed in.
+Then through the `super()` calls, it threads back through the other mixed-in traits. And eventually to the code in 
+`Sample`. Even though none of the traits inherited from one another, it works properly. 
+
+This is similar to the decorator pattern but is more *concise* and less *error-prone*. In other languages, a similar 
+effect could be achieved with a long linear chain of inheritance. But the disadvantage is that for every possible 
+combination of the mix-ins, we need to declare an inheritance chain for it.
+
+
+## Type Enrichment
+Have you ever imagine that you are able to add new function to existing library? An implicit class in Scala can help you 
+to do so. This technique allows new methods to be added to an existing class using an add-on library such that only code 
+that imports the add-on library gets the new functionality, and all other code is unaffected. See the code below:
+```
+object MyExtensions {
+  implicit class IntParity(i: Int) {
+    def isEven = i % 2 == 0
+    def isOdd  = !isEven
+  }
+}
+
+import MyExtensions._  // bring implicit enrichment into scope
+4.isEven  // -> true
+4.isOdd   // -> false
+```
+This is the implicit class that extend the library of class Int. The name of the class does not matter. 
+After importing this class, we can use `isEven()` and `isOdd()` method just as if it has been declared in its own class. 
+
+
+## Pattern Matching
+Scala has built-in support for **pattern matching**, which can be thought of as a more extensible version of a **switch** 
+statement, where arbitrary data types can be matched.
+
+Following is an example of QuickSort algorithm using pattern matching:
+
+```
+def qsort(list: List[Int]): List[Int] = list match {
+  case Nil => Nil
+  case pivot :: tail =>
+    val (smaller, rest) = tail.partition(_ < pivot)
+    qsort(smaller) ::: pivot :: qsort(rest)
+}
+```
+
+The idea here is that we *partition* a list recursively, sort each part and combine the results together.(See [QuickSort](https://en.wikipedia.org/wiki/Quicksort) for the details of the algorithm)
+
+The `match` operator is used to do *pattern matching* on the object stored in list. Each case expression is tried to 
+see if it will match, and the first match will be executed. 
+
+In this case, `Nil` only matches the object `Nil` (Empty list).
+But `pivot :: tail` matches a non-empty list, and it will destructure the list according to the pattern given. 
+In this case, the code will have a variable `pivot` holding the head of the list, and another variable `tail`
+ holding the tail of the list.
+ 
```

Comment:
There are many ways of matching in Scala, case class pattern matching, constructor pattern matching, tuple matching etc, you can either briefly mention them or have a [link](https://alvinalexander.com/scala/how-to-use-pattern-matching-scala-match-case-expressions) to show to the users how strong is Scala's pattern matching.